Mathlib/Combinatorics/SimpleGraph/Matching.lean

```diff
@@ -119,17 +115,104 @@ theorem isPerfectMatching_iff : M.IsPerfectMatching ↔ ∀ v, ∃! w, M.Adj v w
     exact M.edge_vert hw
 #align simple_graph.subgraph.is_perfect_matching_iff SimpleGraph.Subgraph.isPerfectMatching_iff
 
-theorem isPerfectMatching_iff_forall_degree {M : Subgraph G} [∀ v, Fintype (M.neighborSet v)] :
+theorem isPerfectMatching_iff_forall_degree [∀ v, Fintype (M.neighborSet v)] :
     M.IsPerfectMatching ↔ ∀ v, M.degree v = 1 := by
   simp [degree_eq_one_iff_unique_adj, isPerfectMatching_iff]
 #align simple_graph.subgraph.is_perfect_matching_iff_forall_degree SimpleGraph.Subgraph.isPerfectMatching_iff_forall_degree
 
-theorem IsPerfectMatching.even_card {M : Subgraph G} [Fintype V] (h : M.IsPerfectMatching) :
+theorem IsPerfectMatching.even_card [Fintype V] (h : M.IsPerfectMatching) :
     Even (Fintype.card V) := by
   classical
   simpa only [h.2.card_verts] using IsMatching.even_card h.1
 #align simple_graph.subgraph.is_perfect_matching.even_card SimpleGraph.Subgraph.IsPerfectMatching.even_card
 
+lemma IsMatching.induce_connectedComponent (h : M.IsMatching) (c : ConnectedComponent G) :
+    (M.induce (M.verts ∩ c.supp)).IsMatching := by
+  intro v hv
+  simp only [induce_verts, Set.mem_inter_iff, ConnectedComponent.mem_supp_iff] at hv
+  obtain ⟨hv, rfl⟩ := hv
+  obtain ⟨w, hvw, hw⟩ := h hv
+  use w
+  simp only [induce_adj, Set.mem_inter_iff, hv, ConnectedComponent.mem_supp_iff, and_self, and_imp,
+    true_and, ConnectedComponent.eq, hw, hvw, M.edge_vert hvw.symm, (M.adj_sub hvw).symm.reachable]
+  exact fun _ _ _ ↦ hw _
+
+lemma IsPerfectMatching.induce_connectedComponent_isMatching (h : M.IsPerfectMatching)
+    (c : ConnectedComponent G) : (M.induce c.supp).IsMatching := by
+  simpa [h.2.verts_eq_univ] using h.1.induce_connectedComponent c
+
 end Subgraph
 
+namespace ConnectedComponent
+
+section Finite
+
+variable [Fintype V] [DecidableEq V] [DecidableRel G.Adj]
+
+instance instDecidableMemSupp (c : G.ConnectedComponent) (v : V) : Decidable (v ∈ c.supp) :=
+  c.recOn (fun w ↦ decidable_of_iff (G.Reachable v w) $ by simp)
+    (fun _ _ _ _ ↦ Subsingleton.elim _ _)
+
+lemma even_card_of_isPerfectMatching (c : ConnectedComponent G) (hM : M.IsPerfectMatching) :
+    Even (Fintype.card c.supp) := by
+  classical simpa using (hM.induce_connectedComponent_isMatching c).even_card
```

Comment:
`lemma : easy_corollary_of_something_before := by simpa using something_before` is a very good idiom to learn